app/javascript/packages/request/index.ts

```diff
@@ -41,5 +41,9 @@ export async function request<Response>(
   }
 
   const response = await window.fetch(url, { ...fetchOptions, headers, body });
-  return json ? response.json() : response.text();
+  if (response.ok) {
+    return json ? response.json() : response.text();
+  }
+
+  throw new Error(await response.json());
```

Comment:
It'd be good to have spec coverage for the new behaviors here:

- Non-200 status code responses throw
- What happens if the response isn't JSON? I might wonder if it'd be better to avoid using JSON as the error text. Also worth noting that if uncaught, these would be logged to NewRelic, and depending on what's returned from the endpoint JSON, it could include sensitive details we wouldn't want logged (not a problem in this case)

Reply:
>What's the standard approach to this? Is it better to assume `text` because other errors return as text? Some of it is API design — should our API return JSON error bodies? Or text? Etc.

I guess my first question is what details do we want from what's thrown? It doesn't look like we really care what the detail is at the moment, since we're not using it. I think we could even go as far as removing the parameter altogether. But otherwise I think it could be fine to use a general text that's descriptive enough of what happened (e.g. something like `'Request response status was unsuccessful'` 🤷 )

> Also, re: NewRelic, do you mean... the `.json()` parse attempt would fail, uncaught, and log potentially sensitive erro text ("Could not parse token `someone's special info\ `")?

I mean if the API responds with anything other than 200 status code, but sent a JSON payload like...

```
{ ssn: '666-12-1234' }
```

...we'd risk logging that in NewRelic.

I think it can also be helpful to assign a generic name for the error so that we can aggregate it more easily (i.e. otherwise how would we know how often this line of code is throwing?).